notifications-worker.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/dxe/alc-mobile-api/model"
+	"github.com/jmoiron/sqlx"
+	expo "github.com/oliveroneill/exponent-server-sdk-golang/sdk"
+)
+
+func filterValidTokens(notifications []model.Notification) ([]model.Notification, []expo.PushMessage) {
+	var validNotifications []model.Notification
+	var messages []expo.PushMessage
+	for _, n := range notifications {
+		pushToken, err := expo.NewExponentPushToken(n.ExpoPushToken)
+		if err != nil {
+			continue
+		}
+		validNotifications = append(validNotifications, n)
+		messages = append(messages, expo.PushMessage{
+			To:    []expo.ExponentPushToken{pushToken},
+			Title: n.Title,
+			Body:  n.Body,
+		})
+	}
+	return validNotifications, messages
+}
+
+const (
+	StatusSent                = "Sent"
+	StatusDeviceNotRegistered = "DeviceNotRegistered"
+	StatusUnknownError        = "UnknownError"
+)
+
+func notificationsWorker(db *sqlx.DB, client *expo.PushClient) (err error) {
+	notifications, err := model.SelectNotificationsToSend(db)
+	if err != nil {
+		return err
+	}
+
+	if len(notifications) == 0 {
+		return nil
+	}
+
+	validNotifications, validExpoMessages := filterValidTokens(notifications)
+
+	// TODO: use context to cancel if this takes too long?
```

Comment:
I would recommend that. Ideally a context with context.WithDeadline, and then also use the same context for committing the database updates below.

Reply:
Implemented context in a2f1c3e.